### PR TITLE
Vectorized masking

### PR DESCRIPTION
**Pull request by strk**
QGIS enhancement proposal https://github.com/qgis/QGIS-Enhancement-Proposals/issues/186 continuation of PR #38977 after a rebase to current master branch

### CODE REVIEW

**Comment by nyalldawson** [comment]
@strk can you target this for 3.22? If we aim to merge just after thaw we'll have maximum possible time for testing prior to inclusion in a stable release.

**Comment by strk** [comment]
> @strk can you target this for 3.22? If we aim to merge just after thaw we'll have maximum possible time for testing prior to inclusion in a stable release.

What branch targets 3.22 ? I'm targetting the master branch in this PR, cannot find any release_22 branch to target

**Comment by roya0045** [comment]
@strk I think Nyall meant to just wait to merge after 3.20 is released.

**Comment by strk** [comment]
@obrix CI is failing because it doesn't find expected resulting images in the PyQgsSelectiveMasking test, see https://github.com/qgis/QGIS/runs/2600726953#step:12:2366

Missing files:
- highlight_lines.png
- highlight_polygons.png

Did you forget to push those files to your original branch ?

PS: on my system the failure is different, not sure why, I get:
```
661: ======================================================================
661: FAIL: test_layout_exports (__main__.TestSelectiveMasking)
661: Test mask effects in a layout export at 300 dpi
661: ----------------------------------------------------------------------
661: Traceback (most recent call last):
661:   File "/usr/local/src/qgis/qgis/src/smaskingvecto_proto/tests/src/python/test_selective_masking.py", line 700, in test_layout_exports
661:     self.assertTrue(res)
661: AssertionError: False is not true
661: 
661: ----------------------------------------------------------------------
661: Ran 15 tests in 6.498s
```

**Comment by strk** [comment]
@nyalldawson you seem to have changed API a lot, rendering the code in this PR unbuildable. Do you want to help adapting ?

**Comment by nyalldawson** [comment]
@strk do you mean the enum class changes?

**Comment by strk** [comment]
@nyalldawson yes, that one (which I think I can handle) but now I found there's also splitting qgssymbol into per-class translation units... I'll try more of that.

**Comment by strk** [comment]
Great, CI is finally happy \o/

**Comment by nyalldawson** [comment]
@strk can you summarize the rendering loop changes here? (for interested parties without the time available to dig into the code :smile: )

**Comment by obrix** [comment]
@strk About the missing images, I have no idea. The CI was ok last year when the code was ready to be reviewed. Something must have changed. However I remember that the "visual" validation code based on images comparisons was a bit fragile. Thanks a lot to you for taking the time to unearth and update this

@nyalldawson For the rendering part I described everything I did in the qgis enhancement proposal, also interested to know if something in the composition stage was modified by @strk, that was the most painfull part.

**Comment by nyalldawson** [comment]
@obrix I'm mostly interested to know if @strk has done an in-depth review of the changes, or just rebased and fixed the compilation/unit test failures here. I'm still of the opinion that an in-depth review is necessary as the map renderer is critical code. If @strk has done this then I'm happy, but if this branch is just a resurrection of the previous work alone then it still needs a thorough review before it can be merged.

**Comment by wonder-sk** [comment]
FYI - I was asked earlier to do a review, which I am still happy to do... I have asked for:
- addition of auto tests (missing at this point)
- documentation of the approach - from a brief look at the code I quickly got lost what and why is happening. I think we need to have the method properly documented either in the PR description or in the original QEP (which suggest two different solutions, then followed by long discussion difficult to follow). There also seem to be various cases when vector masking falls back to raster masking, such a list is also quite important for future reference.

**Comment by obrix** [comment]
@wonder-sk I haven't looked into qgis for some times, but from what I remember some of the existing selective masking tests were impacted and are actually outputing vectorized test maps. Those outputs are tested against reference images. So there is some tests for the feature. However they can probably be enriched.

For documentation of the approach chosen I think everything is on the qgep thread. Although I can understand this is probably not a friendly topic to go into, necessiting knowledge of the qgis rendering pipeline, qt composition features, and existing selective masking implementation. I was not an expert on those three topics when I did the implementation so clearly a review is needed. Also I think some elements of the approach can still be modified by the review.

**Comment by wonder-sk** [comment]
> some of the existing selective masking tests were impacted and are actually outputing vectorized test maps. Those outputs are tested against reference images. So there is some tests for the feature. However they can probably be enriched.

I can see only one test image changed (+its mask) in the PR. I would expect some tests to be present to verify that the masking is actually vectorized...

As for the modified test image:

image

The rendering output is very complex and difficult to say which image is actually more correct... I have highlighted some differences between the old and the new:

image

Are all these intentional? At least the top-left marked rectangle seems like some unwanted artifact...

> For documentation of the approach chosen I think everything is on the qgep thread. Although I can understand this is probably not a friendly topic to go into, necessiting knowledge of the qgis rendering pipeline, qt composition features, and existing selective masking implementation.

:man_shrugging: I can give it another try, but it is frustrating if documentation that should be in one place is a long thread with a bunch of outdated information... (it doesn't make the review any easier)

**Comment by obrix** [comment]
> The rendering output is very complex and difficult to say which image is actually more correct... I have highlighted some differences between the old and the new

Yes the artifacts are there because of how the vectorized version I implemented tries to also support the transparency and opacity features that was introduced in the rasterized version. This is one of the limitation I referenced in my original PR :

> To support some of the transparency effects that were available in the rasterized version of selective masking, some parts of the rendering process had to stay rasterized. This can lead to the appearance of pixelated area or graphical artefacts when an output vector map is zoomed in.

I think the relevance of the support of those transparency effects and opacity features are questionned in the enhancement proposal. Most of the time there is no artefacts visibles but I did not find any way to avoid some of the artifacts with the solution chosen for the implementation (combination of vectorized elements + rasterized effects).
The "clean" solution would probably be to not support effects on masks at all when outputing vectorized map as vectorized effects are not easily doable with qpainter. But then rasterized outputs and vectorized outputs will not be similar with the same settings.

**Comment by strk** [comment]
> @obrix I'm mostly interested to know if @strk has done an in-depth review of the changes, or just rebased and fixed the compilation/unit test failures here. I'm still of the opinion that an in-depth review is necessary as the map renderer is critical code. If @strk has done this then I'm happy, but if this branch is just a resurrection of the previous work alone then it still needs a thorough review before it can be merged.

I've only done the rebasing and fixing of CI, no deep review.
I don't have much experience with the rendering pipeline of QGIS so I'm not sure what's the best use of my time here.

File: src/core/qgsrendercontext.h [hunk 1]
```diff
@@ -888,6 +912,26 @@ class CORE_EXPORT QgsRenderContext : public QgsTemporalRangeObject
      */
     void setZRange( const QgsDoubleRange &range );
 
+    /**
+     * Add a subpainter for a specific symbolLayer
+     * \since QGIS 3.16
```
Comment (strk) [inline src/core/qgsrendercontext.h:917]: Should all these `\since` indication be updated to read 3.20 ?

Comment (strk) [reply src/core/qgsrendercontext.h:917]: 3.22, actually

File: src/core/maprenderer/qgsmaprendererjob.h [hunk 2]
```diff
@@ -440,7 +459,7 @@ class CORE_EXPORT QgsMapRendererJob : public QObject
      * \note not available in Python bindings
      * \since QGIS 3.12
      */
-    static void composeSecondPass( LayerRenderJobs &secondPassJobs, LabelRenderJob &labelJob ) SIP_SKIP;
+    static void composeSecondPass( LayerRenderJobs &secondPassJobs, LabelRenderJob &labelJob, bool forceVector, bool hasClipping, QPainterPath clipPath ) SIP_SKIP;
```
Comment (strk) [inline src/core/maprenderer/qgsmaprendererjob.h:462]: Should QPainterPath be received by const reference here ?

File: src/core/qgsrendercontext.h [hunk 3]
```diff
@@ -872,6 +873,29 @@ class CORE_EXPORT QgsRenderContext : public QgsTemporalRangeObject
...
+    /**
+     * Add a painter path to the mask painter path of this render context
+     */
+    void addToMaskPainterPath( QPainterPath const &path ); SIP_SKIP
+
+    /**
+     * Returns the mask painter path stored in this render context
+     * \since QGIS 3.22
+     */
+    QPainterPath maskPainterPath() { return mMaskPainterPath; } SIP_SKIP
```
Comment (strk) [inline src/core/qgsrendercontext.h:885]: Should this function return a const reference ?

File: src/core/qgsrendercontext.h [hunk 4]
```diff
@@ -872,6 +873,29 @@ class CORE_EXPORT QgsRenderContext : public QgsTemporalRangeObject
...
+     * Add a painter path to the mask label painter path of this render context with id
+     * \since QGIS 3.22
+     */
+    void addToMaskLabelPainterPath( int id, QPainterPath const &path ); SIP_SKIP
+
+    /**
+     * Returns the label mask painter path with id
+     * \since QGIS 3.22
+     */
+    QPainterPath maskLabelPainterPath( int id = 0 ) { return mMaskLabelPainterPaths[id]; } SIP_SKIP
```
Comment (strk) [inline src/core/qgsrendercontext.h:897]: Should this function return a const reference ?

File: src/core/maprenderer/qgsmaprendererjob.h (hunk 2 above)
Comment (obrix) [reply src/core/maprenderer/qgsmaprendererjob.h:462]: Yes I think it can.

File: src/core/qgsrendercontext.h (hunk 3 above)
Comment (obrix) [reply src/core/qgsrendercontext.h:885]: Also yes.

File: src/core/qgsrendercontext.h (hunk 4 above)
Comment (obrix) [reply src/core/qgsrendercontext.h:897]: Also yes. However I think QPainterPath use implicit sharing in Qt so this a very light object to copy.
